cmake_modules/Configs.cmake

```diff
@@ -0,0 +1,139 @@
+    list(LENGTH defs count)
+    math(EXPR count "${count}-1")
+    foreach(i RANGE ${count})
+        list(GET defs ${i} DEFINITION_KEY)
+        add_definitions(${DEFINITION_KEY}="${defs_map_${DEFINITION_KEY}}")
+    endforeach()
+endmacro()
+
+
+# macro: add_string_config
+#
+# GROUP - the name of the config group
+# KEY - the name of the config key
+# DEFAULT_VALUE - the value to be used for this config in the "default config source"
+# CONFIG_FILE_VALUE - the value to set in the config file, usually this shouild be the same
+#                     as the default.
+#
+# After a call to this macro the following variables are available:
+#   * <GROUP>_<KEY> - set to <DEFAULT_VALUE>
+#   * <GROUP>_<KEY>_CONFIG_FILE_VAR - set to <CONFIG_FILE_VALUE>
+#
+macro(add_string_config GROUP KEY DEFAULT_VALUE CONFIG_FILE_VALUE)
+    set(UNIQUE_KEY "${GROUP}_${KEY}")
+    set(DEFAULT_VALUE_DEFINITION "-D${UNIQUE_KEY}")
+    add_definitions(${DEFAULT_VALUE_DEFINITION}="${DEFAULT_VALUE}")
+
+    set(MANDATORY_DEFINITION "-D${UNIQUE_KEY}_MANDATORY_FLAG")
+
+    # We can't add definitions now since they might be redefined later, so defer the actual
+    # call to 'add_definitions()' to later when we know there will be no more changes to
+    # definitons. A call to 'set_all_deferred_config_definitions()' will trigger the deferred adding
+    # of definitions.
+    defer_add_definitions(${MANDATORY_DEFINITION} FALSE)
+
+    # Set the key so it's available for other exports and uses of the value in the cmake files.
+    set(${UNIQUE_KEY} "${DEFAULT_VALUE}")
+
+    set("${UNIQUE_KEY}_ACTIVATE" "# ")
+    set("${UNIQUE_KEY}_CONFIG_FILE_VAR" "${CONFIG_FILE_VALUE}")
+endmacro()
+
+
+# Same as above but without the quoted values (which are not needed for non string values).
+macro(add_integer_config GROUP KEY DEFAULT_VALUE CONFIG_FILE_VALUE)
+    set(UNIQUE_KEY "${GROUP}_${KEY}")
+    set(DEFAULT_VALUE_DEFINITION "-D${UNIQUE_KEY}")
+    add_definitions(${DEFAULT_VALUE_DEFINITION}=${DEFAULT_VALUE})
+
+    set(MANDATORY_DEFINITION "-D${UNIQUE_KEY}_MANDATORY_FLAG")
+
+    # We can't add definitions now since they might be redefined later, so defer the actual
+    # call to 'add_definitions()' to later when we know there will be no more changes to
+    # definitons. A call to 'set_all_deferred_config_definitions()' will trigger the deferred adding
+    # of definitions.
+    defer_add_definitions(${MANDATORY_DEFINITION} FALSE)
+
+    # Set the key so it's available for other exports and uses of the value in the cmake files.
+    set(${UNIQUE_KEY} ${DEFAULT_VALUE})
+
+    set("${UNIQUE_KEY}_ACTIVATE" "# ")
+    set("${UNIQUE_KEY}_CONFIG_FILE_VAR" ${CONFIG_FILE_VALUE})
+endmacro()
+
+
+# The boolean and integer configs are added in the same way, so this macro just
+# passes along the args to add_boolean_config().
+macro(add_boolean_config GROUP KEY DEFAULT_VALUE CONFIG_FILE_VALUE)
+    add_integer_config(${GROUP} ${KEY} ${DEFAULT_VALUE} ${CONFIG_FILE_VALUE})
```

Comment:
[Copied from CMakeList.txt] How come the "default value" is (cmake) bool and "config file value" a string?